src/lib.rs

```diff
@@ -42,58 +43,63 @@ impl Parse for Generic {
+                Ok(type_) => Ok(Self::NoBound(type_)),
+                Err(error) => Err(Error::new(error.span(), &format!("expected type to bind to, {}", error))),
+            },
         }
     }
 }
 
 /// Holds parsed [generics](Generic) and [traits](Trait).
 struct DeriveWhere {
-    /// [generics](Generic) for where clause
-    generics: Option<Vec<Generic>>,
-    /// [traits](Trait) to implement
+    /// [traits](Trait) to implement.
     traits: Vec<Trait>,
+    /// [generics](Generic) for where clause.
+    generics: Option<Vec<Generic>>,
 }
 
 impl Parse for DeriveWhere {
-    /// Parse the macro input this should either be:
-    /// - Comma seperated traits
-    /// - Comma seperated generics `;` comma sperated traits
+    /// Parse the macro input, this should either be:
+    /// - Comma separated traits
+    /// - Comma separated traits `;` Comma separated generics
     fn parse(input: ParseStream) -> syn::Result<Self> {
-        let fork = input.fork();
-        // Try to parse input as only a trait list. This should fail fast due
-        // to trait names not commonly being used as generic parameters.
-        match Punctuated::<Trait, Token![,]>::parse_terminated(&fork) {
-            Ok(derive_where) => {
-                // Advance input as if `DeriveWhere` was parsed on it.
-                input.advance_to(&fork);
-                Ok(Self {
-                    generics: None,
-                    traits: derive_where.into_iter().collect(),
-                })
-            }
-            Err(_) => {
-                let generics = Some(
-                    Punctuated::<Generic, Token![,]>::parse_separated_nonempty(input)?
-                        .into_iter()
-                        .collect(),
-                );
-                <Token![;]>::parse(input)?;
-                let traits = Punctuated::<Trait, Token![,]>::parse_terminated(input)?
-                    .into_iter()
-                    .collect();
-
-                Ok(Self { generics, traits })
+        let mut traits = Vec::new();
+        let mut generics = None;
+
+        // Start parsing traits.
+        while !input.is_empty() {
+            traits.push(Trait::parse(input)?);
+
+            if !input.is_empty() {
+                let fork = input.fork();
+
+                if <Token![;]>::parse(&fork).is_ok() {
+                    input.advance_to(&fork);
+
+                    // If we found a semi-colon, start parsing generics.
+                    if !input.is_empty() {
+                        generics = Some(
+                            Punctuated::<Generic, Token![,]>::parse_terminated(input)?
+                                .into_iter()
+                                .collect(),
+                        );
+                    }
+                } else if let Err(error) = <Token![,]>::parse(input) {
+                    return Err(Error::new(error.span(), "expected `;` or `,"));
+                }
```

Comment:
:+1: manual parsing and yet almost more readable than my solution

Reply:
Uff, the first time around when I swapped the syntax without manual parsing it was so beautiful, this looks super-messy to me now :cry:.